### PR TITLE
POST: Year of Hell

### PR DESCRIPTION
**Pull request by klaussfreire**
A retrospective look at the rewrite project we called Silver,
from a Core-centric PoV, with lessons learned and a lot of
sugar coating.

It may need a ton of beautification.

### CODE REVIEW

File: _src/_posts/2019-02-26-year-of-hell.markdown [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+No way José. I've done those, I have developed the knee-jerk self-preservation
+reflex to say "No" when I hear such proposals. No, we didn't **intend** to do a rewrite.
+
+We just ended up doing it.
+
+## Preface
+
+Let me tell you the story of how we ended up there in the first place.
+If you're feeling like "TL;DR, gimme the punchline", feel free to skip to the
+next section. But if you're in the mood for some storytelling, read on.
```
Comment (klaussfreire) [inline _src/_posts/2019-02-26-year-of-hell.markdown]: Alguien sabe cómo poner un link ahí a la sección siguiente? (un anchor)

File: _src/_posts/2019-02-26-year-of-hell.markdown [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+And if you've been in that situation and decided to go for it, [you know
+how it ends as well](https://medium.com/@herbcaudill/lessons-from-6-software-rewrite-stories-635e4c8f7c22).
+Almost always, in flames.
+<!--excerpt.end-->
+
+Rewrites are very difficult, very risky projects:
+
+* **They're extremely long projects**
+  
+  It's not due to laziness
```
Comment (tzulberti-jampp) [inline _src/_posts/2019-02-26-year-of-hell.markdown]: Isn't this missing a `,` (comma)?

File: _src/_posts/2019-02-26-year-of-hell.markdown [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+
+Instead of processing campaigns one by one, caching results as we did
+before, we precomputed a decision tree optimized for whatever the current
+campaign setup is at the time. A greedy algorithm would be free to
+shape that BDT the way it best suited the current setup, making it a very
+flexible design.
+
+The decision tree would trade offline CPU time with online CPU time.
+Building the tree would take a lot of CPU time, but the structure can
+then be shared by all bidders and that cost gets amortized in time and scale.
```
Comment (tzulberti-jampp) [inline _src/_posts/2019-02-26-year-of-hell.markdown]: `bidder instances` (or something like that).

File: _src/_posts/2019-02-26-year-of-hell.markdown (hunk 2 above)
Comment (klaussfreire) [reply _src/_posts/2019-02-26-year-of-hell.markdown]: Where exactly?

File: _src/_posts/2019-02-26-year-of-hell.markdown (hunk 3 above)
Comment (klaussfreire) [reply _src/_posts/2019-02-26-year-of-hell.markdown]: I don't think it adds much to outsiders. It may even be confusing for them.

**Pull request merged**